### DISCUSSION TITLE
Clarification on constrained planning constraints for non axis aligned movements

### DISCUSSION THREAD

**Discussion by blooop** [NVlabs/curobo#206] [Software Q&A] [answered]
Thanks for adding the constrained planning feature!

I have created an example based off the constrained planning example to match my needs but would like some clarification on how specify the constraints.

I have 3 points in a right angle triangle on the groundplane.
start: (0,0,0)
p1: (1,0,0)
p2: (1,1,0)

to move from start -> p1 I have the robot constrained in orientation, z and move along the x direction and use hold_vec_weight= [1,1,1,1,0,1]
to move from p1 ->p2 I constrain in orientation, z and move along y with hold_vec_weight= [1,1,1,0,1,1]

I would like to move back to the start point but can't find the correct hold_vec_weight to do this.

I have tried:
[1, 1, 1, 0.5, 0.5, 1]
[1, 1, 1, 1.4142, 1.4142, 0]
[1, 1, 1, 1.4142, 1.4142, 1]
[1, 1, 1, 0.70710, 0.70710, 0]
[1, 1, 1, 0.70710, 0.70710, 1]

How would I go about constraining the pose of the end effector in a non axis aligned way?

Thanks

**Comment by balakumar-s** [comment]
We should be able to support this. There are two frame of references that can be used when calculating the pose cost. The first one is using the robot's base frame and second one being the target pose frame. Currently, we have to choose the reference frame during initialization of `MotionGen` which is not ideal. We are working on making this a dynamic parameter so that one could change the frame of reference between planning calls. This could handle non-axis aligned constrained planning.

In the meantime, you can get close to what you want by doing the following:

1. From your description, it looks like you have the frame of reference for pose cost set as the goal frame. To use base frame as the frame of reference for pose cost, set this to `False`: https://github.com/NVlabs/curobo/blob/b481ee201a2754c072ba9bcd80bd5d327fddc599/src/curobo/wrap/reacher/motion_gen.py#L207 

2. Change the hold_vec_weight to below:

| Name |Goal position | hold_vec_weight | Required Behavior |
|-------|--------------|------------------|-------------------|
| start -> p1    | (0,0,0) -> (1,0,0) |  [1,1,1,0,1,1] | robot moves only along x (reference frame=base) while maintaining orientation|
| p1  -> p2      | (1,0,0) -> (1,1,0)  |  [1,1,1,1,0,1] | robot moves only along y (reference frame=base) while maintaining orientation|
| p2  -> start  | (1,1,0) -> (0,0,0)  | [1,1,1,0,0,1] | robot moves only along x-y (reference frame=base) while maintaining orientation |

For reference, `hold_vec_weight=[orientation_x, orientation_y, orientation_z, linear_x, linear_y, linear_z]`

Note that when moving from p2 -> start, the robot is constrained only on z, so the path taken in x-y plane might be not a linear path in Cartesian Space. Once we have the dynamic frame change implemented, we could constrain the p2 -> start to be a linear path in Cartesian space.

**Reply by blooop** [reply]
Thanks for your detailed reply.

For my use case I would need it to be a linear path from p2->start as well so I don't think the workaround will work, but as another potential workaround, for p2->start could I do something like:

Rotate the end effector by 45 degrees so that it is axis aligned with the direction I want to go and use

hold_vec_weight = [1,1,1,0,1,1] (reference_frame=end effector)

That should give a linear path in Cartesian space in the direction I want to move. 

I was hoping to write a wrapper that allowed linear Cartesian motion from any xyz point to any other xyz point while maintaining orientation and additionally generating paths with constrained Cartesian motion and spherical linear interpolated orientation, but it sounds that that would be a lot harder to support.

**Reply by balakumar-s** [reply, marked as answer]
> Rotate the end effector by 45 degrees so that it is axis aligned with the direction I want to go and use

Yes, this would work.


> spherical linear interpolated orientation

This is not possible currently and also might be hard to implement.


Also, we are thinking of having two flags for reference frame in pose cost, one for linear axes and other for angular. That way you can maintain orientation w.r.t. base frame while maintain/move linearly w.r.t. target frame.

**Reply by blooop** [reply]
> Also, we are thinking of having two flags for reference frame in pose cost, one for linear axes and other for angular. That way you can maintain orientation w.r.t. base frame while maintain/move linearly w.r.t. target frame.

This sounds like a useful feature.
